File: bot/leaderboard.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from datetime import date, datetime

_PER_EMBED = 50


@dataclass
class PlayerTotal:
    player_key: str
    display_name: str
    points: int
    events: int


def month_window(now: datetime) -> tuple[date, date]:
    today = now.date()
    return today.replace(day=1), today


def aggregate_totals(rows: list[dict]) -> list[PlayerTotal]:
    acc: dict[str, PlayerTotal] = {}
    for row in rows:
        key = row["player_key"]
        current = acc.get(key)
        if current is None:
            acc[key] = PlayerTotal(key, row["player_name"], row["points"], 1)
        else:
            # rows arrive oldest-first, so overwrite display name with the latest
            acc[key] = PlayerTotal(
                key,
                row["player_name"],
                current.points + row["points"],
                current.events + 1,
            )
    return sorted(acc.values(), key=lambda t: (-t.points, t.display_name.lower()))


def _line(rank: int, t: PlayerTotal) -> str:
    unit = "event" if t.events == 1 else "events"
    return f"{rank}. {t.display_name} — {t.points} pts ({t.events} {unit})"


def build_leaderboard_embeds(totals: list[PlayerTotal], month_label: str) -> list[dict]:
    embeds: list[dict] = []
    for start in range(0, max(len(totals), 1), _PER_EMBED):
        chunk = totals[start : start + _PER_EMBED]
        lines = [_line(start + i + 1, t) for i, t in enumerate(chunk)]
        title = f"\U0001f3c6 Standard League — {month_label}"
        if start > 0:
            title += " (cont.)"
        embeds.append({"title": title, "description": "\n".join(lines)})
    return embeds


_MONTHS = [
    "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
]


def _is_summer_2026(event_date) -> bool:
    if isinstance(event_date, str):
        year, month = int(event_date[:4]), int(event_date[5:7])
    else:
        year, month = event_date.year, event_date.month
    return year == 2026 and month in (6, 7, 8)


def _ranked(rows: list[dict]) -> list[dict]:
    """Order a tournament's rows by finishing position.

    Standings-shaped events store the organiser's final ranking in `pairing`,
    one row per player. That ranking already accounts for tiebreakers we do not
    store (OMW%/GW%/OGW%), so recomputing an order here would drop them and fall
    back to alphabetical. Prefer the stored ranking when it is one.

    In pairing-shaped events two players share a `pairing` — it is a table
    number, not a rank — so derive the order from the records instead.
    """
    pairings = [r.get("pairing") for r in rows]
    if all(p is not None for p in pairings) and len(set(pairings)) == len(rows):
        return sorted(rows, key=lambda r: r["pairing"])
    return sorted(
        rows,
        key=lambda r: (
            -(3 * r["record_wins"] + r["record_draws"]),
            -(r["game_wins"] or 0),
            r["player_name"].lower(),
        ),
    )


def _tournament_scores(rows: list[dict]) -> dict:
    summer = _is_summer_2026(rows[0]["event_date"])
    ranked = _ranked(rows)
    bonus = [3, 2, 1]
    out: dict = {}
    for i, r in enumerate(ranked):
        if summer:
            placement = bonus[i] if i < 3 else 0
            score = placement + 2 * r["record_wins"] + r["record_draws"] + 1
        else:
            score = 3 * r["record_wins"] + r["record_draws"]
        out[r["player_key"]] = {"player_name": r["player_name"], "score": score}
    return out


def season_totals(stats: list[dict], league_keys: set[str]) -> list[PlayerTotal]:
    by_tournament: dict = defaultdict(list)
    for row in stats:
        by_tournament[row["tournament_id"]].append(row)
    acc: dict[str, PlayerTotal] = {}
    for rows in by_tournament.values():
        for key, s in _tournament_scores(rows).items():
            if key not in league_keys:
                continue
            cur = acc.get(key)
            if cur is None:
                acc[key] = PlayerTotal(key, s["player_name"], s["score"], 1)
            else:
                acc[key] = PlayerTotal(
                    key, s["player_name"], cur.points + s["score"], cur.events + 1
                )
    return sorted(acc.values(), key=lambda t: (-t.points, t.display_name.lower()))


def run(discord, store, channel_id: str, now: datetime) -> bool:
    start, end = month_window(now)
    stats = store.fetch_tournament_stats(start, end)
    league_keys = store.fetch_league_keys()
    totals = season_totals(stats, league_keys)
    if not totals:
        return False
    label = f"{_MONTHS[start.month - 1]} {start.year}"
    embeds = build_leaderboard_embeds(totals, label)
    discord.post_embeds(channel_id, embeds)
    return True

File: web/lib/leaderboard.js
import { seasonKey } from './season.js';

function playerEntry(player) {
  return { record: player.record, isLeague: player.is_league !== false };
}

export function finalRecords(tournament) {
  const last = {};
  for (const round of tournament.rounds) {
    for (const pairing of round.pairings) {
      last[pairing.player1.name] = playerEntry(pairing.player1);
      if (pairing.player2) {
        last[pairing.player2.name] = playerEntry(pairing.player2);
      }
    }
  }
  return last;
}

export function points(record) {
  return record.wins * 3 + record.draws;
}

function accumulate(stats, player, pairing) {
  const s = stats[player.name] || { record: null, gameWins: 0, isLeague: true };
  s.record = player.record; // rounds are in order, so this ends as the final record
  s.gameWins += player.game_wins || 0;
  s.isLeague = player.is_league !== false;
  s.pairing = pairing; // rounds are in order, so this ends as the final pairing
  stats[player.name] = s;
}

export function playerTournamentStats(tournament) {
  const stats = {};
  for (const round of tournament.rounds) {
    for (const pairing of round.pairings) {
      accumulate(stats, pairing.player1, pairing.pairing);
      if (pairing.player2) accumulate(stats, pairing.player2, pairing.pairing);
    }
  }
  return stats;
}

const ORD = ['1st', '2nd', '3rd'];

function plural(n, word) {
  return `${n} ${n === 1 ? word : word + 's'}`;
}

// Standings-shaped events store the organiser's final ranking in `pairing`, one
// player per pairing. That ranking already accounts for tiebreakers the data does
// not carry (OMW%/GW%/OGW%), so it is authoritative. In pairing-shaped events two
// players share a `pairing` — it is a table number, not a rank — so the order has
// to be derived from the records instead.
function rankPlayers(players) {
  const pairings = players.map(p => p.pairing);
  const isStandings =
    pairings.every(p => p !== undefined && p !== null) &&
    new Set(pairings).size === players.length;
  if (isStandings) return [...players].sort((a, b) => a.pairing - b.pairing);
  return [...players].sort(
    (a, b) => b.mp - a.mp || b.gameWins - a.gameWins || a.name.localeCompare(b.name),
  );
}

export function tournamentScores(tournament) {
  const stats = playerTournamentStats(tournament);
  const summer = seasonKey(tournament.date) === '2026-2';
  const ranked = rankPlayers(
    Object.entries(stats).map(([name, s]) => ({ name, ...s, mp: points(s.record) })),
  );
  const bonus = [3, 2, 1];
  const scores = {};
  ranked.forEach((p, i) => {
    const items = [];
    if (summer) {
      if (i < 3) items.push({ label: `${ORD[i]} place`, points: bonus[i] });
      if (p.record.wins > 0) items.push({ label: `${plural(p.record.wins, 'win')} (×2)`, points: 2 * p.record.wins });
      if (p.record.draws > 0) items.push({ label: `${plural(p.record.draws, 'draw')} (×1)`, points: p.record.draws });
      items.push({ label: 'attendance', points: 1 });
    } else {
      if (p.record.wins > 0) items.push({ label: `${plural(p.record.wins, 'win')} (×3)`, points: 3 * p.record.wins });
      if (p.record.draws > 0) items.push({ label: `${plural(p.record.draws, 'draw')} (×1)`, points: p.record.draws });
    }
    const subtotal = items.reduce((sum, it) => sum + it.points, 0);
    scores[p.name] = {
      score: subtotal,
      isLeague: p.isLeague,
      breakdown: { tournament: tournament.name, date: tournament.date, items, subtotal },
    };
  });
  return scores;
}

function buildBoard(tournaments) {
  const agg = {};
  for (const tournament of tournaments) {
    const scored = tournamentScores(tournament);
    for (const [name, { score, isLeague, breakdown }] of Object.entries(scored)) {
      if (!isLeague) continue;
      if (!agg[name]) agg[name] = { name, points: 0, events: 0, breakdown: [] };
      agg[name].points += score;
      agg[name].events += 1;
      agg[name].breakdown.push(breakdown);
    }
  }
  return Object.values(agg).sort(
    (a, b) => b.points - a.points || a.name.localeCompare(b.name),
  );
}

export function seasonLeaderboard(tournaments, key) {
  return buildBoard(tournaments.filter(t => seasonKey(t.date) === key));
}

// The season board as it stood before the latest-dated tournament(s) in the
// season, for computing position movement. `hasPrevious` is false when the
// season has no earlier tournament to compare against.
export function seasonLeaderboardBefore(tournaments, key) {
  const inSeason = tournaments.filter(t => seasonKey(t.date) === key);
  if (inSeason.length === 0) return { rows: [], hasPrevious: false };
  const maxDate = inSeason.reduce((m, t) => (t.date > m ? t.date : m), inSeason[0].date);
  const earlier = inSeason.filter(t => t.date < maxDate);
  return { rows: buildBoard(earlier), hasPrevious: earlier.length > 0 };
}
